### PR TITLE
Ensure that duplicates don't end up in dev mode CP in multi-module projects

### LINKED ISSUE #10411: Devmode, Liquibase: Changelog file is found twice if located in another Maven module
**Describe the bug**
Application startup fails in Devmode if I try to auto-migrate with a changelog file that comes from another Maven module because Liquibase finds it two times on the classpath (once in the jar and once from the module `target`).

**Expected behavior**
No failure. In Devmode, the one from `target` should be used, IMHO.

**Actual behavior**
```
ERROR: Failed to start application
java.lang.IllegalStateException: liquibase.exception.ChangeLogParseException: Error Reading Migration File: Found 2 files that match db/changeLog.xml
        at io.quarkus.liquibase.runtime.LiquibaseRecorder.doStartActions(LiquibaseRecorder.java:69)
        at io.quarkus.deployment.steps.LiquibaseProcessor$createBeansAndStartActions2123523327.deploy_0(LiquibaseProcessor$createBeansAndStartActions2123523327.zig:76)
        at io.quarkus.deployment.steps.LiquibaseProcessor$createBeansAndStartActions2123523327.deploy(LiquibaseProcessor$createBeansAndStartActions2123523327.zig:36)
        at io.quarkus.runner.ApplicationImpl.doStart(ApplicationImpl.zig:356)
        at io.quarkus.runtime.Application.start(Application.java:90)
        at io.quarkus.runtime.ApplicationLifecycleManager.run(ApplicationLifecycleManager.java:90)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:61)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:38)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:106)
        at io.quarkus.runner.GeneratedMain.main(GeneratedMain.zig:29)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at io.quarkus.runner.bootstrap.StartupActionImpl$3.run(StartupActionImpl.java:144)
        at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: liquibase.exception.ChangeLogParseException: Error Reading Migration File: Found 2 files that match db/changeLog.xml
        at liquibase.parser.core.xml.XMLChangeLogSAXParser.parseToNode(XMLChangeLogSAXParser.java:118)
        at liquibase.parser.core.xml.AbstractChangeLogParser.parse(AbstractChangeLogParser.java:15)
        at liquibase.Liquibase.getDatabaseChangeLog(Liquibase.java:217)
        at liquibase.Liquibase.validate(Liquibase.java:1559)
        at io.quarkus.liquibase.runtime.LiquibaseRecorder.doStartActions(LiquibaseRecorder.java:60)
        ... 15 more
Caused by: java.io.IOException: Found 2 files that match db/changeLog.xml
        at liquibase.util.StreamUtil.singleInputStream(StreamUtil.java:206)
        at liquibase.parser.core.xml.XMLChangeLogSAXParser.parseToNode(XMLChangeLogSAXParser.java:71)
        ... 19 more
```

**To Reproduce**
1. Clone https://github.com/famod/quarkus-liquibase-dup
    (this is just the official quickstart split into two modules and without h2 tcp, which caused problems on my machine)
2. Passes: `mvn clean install` (includes a test)
3. Passes: `java -jar app/target/liquibase-dup-app-1.0-SNAPSHOT-runner.jar`
4. Fails: `mvn quarkus:dev -f app` (or without `-f` from within `app` )

**Configuration**
```properties
quarkus.datasource.db-kind=h2
quarkus.datasource.username=sa
quarkus.datasource.password=sa
quarkus.datasource.jdbc.url=jdbc:h2:mem:test_quarkus;DB_CLOSE_DELAY=-1
quarkus.liquibase.migrate-at-start=true
```

**Screenshots**
n/a

**Environment (please complete the following information):**
 - Output of `uname -a` or `ver`:
    ```
    MINGW64_NT-10.0-18363 XXX 3.1.4-340.x86_64 2020-05-19 12:55 UTC x86_64 Msys
    ```
 - Output of `java -version`:
    ```
    openjdk version "11.0.6" 2020-01-14
    OpenJDK Runtime Environment AdoptOpenJDK (build 11.0.6+10)
    OpenJDK 64-Bit Server VM AdoptOpenJDK (build 11.0.6+10, mixed mode)
    ```
 - GraalVM version (if different from Java): n/a
 - Quarkus version or git rev: `1.4.2.Final`, `1.5.2.Final`, `1.6.0.CR1`
 - Build tool (ie. output of `mvnw --version` or `gradlew --version`): `Apache Maven 3.6.3`

**Additional context**
There have been two fixes before, but both adressed Gradle:
- #8400
- #8984

### PR DESCRIPTION
**Pull request by geoand**
Fixes: #10411

### CODE REVIEW

**Review by aloubyansky — APPROVED** [review]
Looks good!

File: devtools/maven/src/main/java/io/quarkus/maven/DevMojo.java [hunk 1]
```diff
@@ -657,7 +657,12 @@ void prepare(final boolean triggerCompile) throws Exception {
             //in most cases these are not used, however they need to be present for some
             //parent-first cases such as logging
             for (Artifact appDep : project.getArtifacts()) {
-                addToClassPaths(classPathManifest, appDep.getFile());
+                // only add the artifact if it's present in the dev mode context
+                // we need this to avoid having jars on the classpath multiple times
+                if (!devModeContext.getLocalArtifacts().contains(new AppArtifactKey(appDep.getGroupId(), appDep.getArtifactId(),
+                        appDep.getClassifier(), appDep.getType()))) {
```
Comment (aloubyansky) [inline devtools/maven/src/main/java/io/quarkus/maven/DevMojo.java:663]: appDep.getArtifactHandler().getExtension() perhaps would be a more accurate one.

Comment (geoand) [reply devtools/maven/src/main/java/io/quarkus/maven/DevMojo.java]: Didn't know that one, thanks!

**Comment by geoand** [comment]
This is a good candidate for 1.6.1

**Comment by famod** [comment]
@geoand 

> This is a good candidate for 1.6.1, hence the backport label

Just asking for sprint planning reasons: 1.6.0 is already "frozen"?

**Comment by geoand** [comment]
> @geoand
> 
> > This is a good candidate for 1.6.1, hence the backport label
> 
> Just asking for sprint planning reasons: 1.6.0 is already "frozen"?

Yup. We will only add bug fixes to 1.6.x releases (if needed)

**Comment by famod** [comment]
Now that's interesting and a bit confusing: In context of [my CCE-problem with CXF](https://github.com/quarkusio/quarkus/issues/10411#issuecomment-652592387), I tried out this PR branch **and the behaviour changes**: I now get a `ClassNotFoundException` instead of `ClassCastException` (for the very same class).

Runner-jar is still fine. 😕 

I guess it is time to create a separate issue, I guess?

**Comment by aloubyansky** [comment]
@famod could this be reproduced with the same reproducer referenced from #10411 ?

**Comment by geoand** [comment]
> Now that's interesting and a bit confusing: In context of [my CCE-problem with CXF](https://github.com/quarkusio/quarkus/issues/10411#issuecomment-652592387), I tried out this PR branch **and the behaviour changes**: I now get a `ClassNotFoundException` instead of `ClassCastException` (for the very same class).
> 
> Runner-jar is still fine. 
> 
> I guess it is time to create a separate issue?

Yup

**Comment by famod** [comment]
@aloubyansky Unfortunately not. It seems it really requires CXF which makes it harder to reproduce.
I'll try to come up with a reproducer and a new issue then.

Apart from that: This PR _does_ fix #10411 for my internal project! Good work, thanks! Looking forward to 1.6.1.

**Comment by geoand** [comment]
Darn CI failed... It was passing locally... So now I have to figure out what's going on

**Comment by aloubyansky** [comment]
Wait for the CI to finish, we'll check the logs.

**Comment by geoand** [comment]
> Wait for the CI to finish, we'll check the logs.

I'll take care of it, don't worry :)

**Comment by geoand** [comment]
CI passed on my fork, so let's see what happens here

**Pull request merged**